arroyo/processing/strategies/dead_letter_queue/policies/produce.py

```diff
@@ -0,0 +1,89 @@
+import base64
+import json
+import time
+from collections import deque
+from concurrent.futures import Future
+from typing import Deque, Optional
+
+from arroyo.backends.kafka.consumer import KafkaPayload, KafkaProducer
+from arroyo.processing.strategies.dead_letter_queue.policies.abstract import (
+    DeadLetterQueuePolicy,
+    InvalidMessages,
+    Serializable,
+)
+from arroyo.types import Message, Topic
+from arroyo.utils.metrics import get_metrics
+
+MAX_QUEUE_SIZE = 5000
+
+
+class ProduceInvalidMessagePolicy(DeadLetterQueuePolicy):
+    """
+    Produces given InvalidMessages to a dead letter topic.
+    """
+
+    def __init__(self, producer: KafkaProducer, dead_letter_topic: Topic) -> None:
+        self.__metrics = get_metrics()
+        self.__dead_letter_topic = dead_letter_topic
+        self.__producer = producer
+        self.__futures: Deque[Future[Message[KafkaPayload]]] = deque()
+
+    def handle_invalid_messages(self, e: InvalidMessages) -> None:
+        """
+        Produces a message to the given dead letter topic for each
+        invalid message in the form:
+
+        {
+            "topic": <original topic the bad message was produced to>,
+            "reason": <why the message(s) are bad>
+            "timestamp": <time at which exception was thrown>,
+            "message": <original bad message>
+        }
+        """
+        for message in e.messages:
+            payload = self._build_payload(e, message)
+            self._produce(payload)
+        self.__metrics.increment("dlq.produced_messages", len(e.messages))
+
+    def _build_payload(self, e: InvalidMessages, message: Serializable) -> KafkaPayload:
+        if isinstance(message, bytes):
+            message = base64.b64encode(message).decode("utf-8")
```

Comment:
JSON serializer needs strings or other serializable types. `bytes` is not JSON serializable, so base 64 encodes and then UTF-8 decodes back into a string to be placed in the payload.

@fpacifici 
Why is base 64 encoding needed, could we not directly decode `bytes` into a string like `message.decode("utf-8")` to be placed in the payload?

Reply:
No because the bytes string you have may not be a valid UTF-8 encoded string. Which can be one of the reason why the message was invalid 
```
>>> a = b'\xff'
>>> a
b'\xff'
>>> a.decode("utf-8")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
```

Reply:
You can try to decode and see if it is a valid string, then add the string. and encode base64 if the bytes string is not a valid UTF-8 string.

Reply:
Fixed and moved this logic into the `InvalidMessage` dataclass.